### PR TITLE
[TypeScript SDK] Derive Mutable for transactions that touches shared objects

### PR DESCRIPTION
**Pull request by healthydeve**
## Description 

Describe the changes or additions included in this PR.

## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: sdk/typescript/src/types/sui-bcs.ts [hunk 1]
```diff
@@ -24,80 +24,6 @@ function registerUTF8String(bcs: BCS) {
   );
 }
 
-/**
```
Comment (666lcz) [inline sdk/typescript/src/types/sui-bcs.ts:27]: This is due to the merge with https://github.com/MystenLabs/sui/pull/9098 and https://github.com/MystenLabs/sui/pull/9101. Once these PRs lands, we can do a rebase

File: sdk/typescript/src/builder/Transaction.ts [hunk 2]
```diff
@@ -419,11 +438,13 @@ export class Transaction {
...
       objects.forEach((object, i) => {
-        const { id, input } = objectsToResolve[i];
+        const { id, input, normalizedType } = objectsToResolve[i];
         const initialSharedVersion = getSharedObjectInitialVersion(object);
 
         if (initialSharedVersion) {
-          const mutable = true; // Defaulted to True to match current behavior.
+          const mutable =
+            normalizedType != null &&
+            extractMutableReference(normalizedType) != null;
```
Comment (666lcz) [inline sdk/typescript/src/builder/Transaction.ts:447]: cc @amnn this is the latest set mutable change

File: crates/sui-transaction-builder/src/lib.rs [hunk 3]
```diff
@@ -356,7 +359,11 @@ impl<Mode: ExecutionMode> TransactionBuilder<Mode> {
                 id,
                 initial_shared_version,
                 // todo(RWLocks) - do we want to parametrise this?
-                mutable: true, // using mutable reference by default here.
+                mutable: match expected_type {
+                    // Only set mutable reference types to true
+                    SignatureToken::MutableReference(_) => true,
+                    _ => false,
+                },
```
Comment (amnn) [inline crates/sui-transaction-builder/src/lib.rs:366]: nit
```suggestion
                mutable: matches!(expected_type, SignatureToken::MutableReference(_)),
```

File: sdk/typescript/src/builder/Transaction.ts (hunk 2 above)
Comment (amnn) [reply sdk/typescript/src/builder/Transaction.ts:447]: Great, thanks for already being on top of this!  qq -- what happens if the same Object ID is required by multiple calls in the transaction, potentially with different type signatures?  I would expect to see logic to de-duplicate by ID, and pick an upperbound type for the transaction input based on the call parameter types (by-value beats by-mutable-reference, which beats by-reference) but I didn't spot that.  Is that scenario even supported?

Comment (666lcz) [reply sdk/typescript/src/builder/Transaction.ts:447]: NIce catch, I'll fix that!

File: crates/sui-transaction-builder/src/lib.rs [hunk 4]
```diff
@@ -400,16 +402,19 @@ impl<Mode: ExecutionMode> TransactionBuilder<Mode> {
...
-                }
+                SuiJsonCallArg::Object(id) => CheckCallArg::Object(
+                    self.get_object_arg(id, &mut objects, expected_type).await?,
+                ),
                 SuiJsonCallArg::Pure(p) => CheckCallArg::Pure(p),
                 SuiJsonCallArg::ObjVec(v) => {
                     let mut object_ids = vec![];
                     for id in v {
-                        object_ids.push(self.get_object_arg(id, &mut objects).await?);
+                        object_ids.push(
```
Comment (666lcz) [inline crates/sui-transaction-builder/src/lib.rs:414]: There are some edge case here: https://mysten-labs.slack.com/archives/C04RHGZ75A4/p1678490907579999?thread_ts=1678462995.547919&cid=C04RHGZ75A4 will follow up in a separate PR

**Review by 666lcz — APPROVED** [review]
There are still some edge cases as mentioned inline to be handled. Will merge this now and tackle them as a separate PR so that it's easier to review and rebase

File: crates/sui-transaction-builder/src/lib.rs (hunk 4 above)
Comment (666lcz) [reply crates/sui-transaction-builder/src/lib.rs:414]: Resolved by https://github.com/MystenLabs/sui/pull/9130

**Pull request merged**